sql/schemachanger: fix bugs with the STORING() clause of CREATE INDEX.

spilchen opened the pull request:
Issue #124511 uncovered two problems with the Declarative Schema Changer (DSC):

1. If a column name contains a UTF-8 character and is included in the STORING() clause, the DSC cannot detect that it duplicates the primary key.
2. If a virtual column is included in the STORING() clause, the DSC does not catch this upfront, unlike the legacy schema changer, leading to an internal error.

The first problem occurs because we build the list of columns referenced in an index by their names. These names are normalized to handle UTF-8 encoding. However, when checking if a storing column is already referenced, we weren't using the normalized version of the string, resulting in a mismatch.

The second problem arises from inadequate checks to detect if a virtual column is included in the STORING() clause. Adding a check for this resolves the issue.

Fixes #124511
Fixes https://github.com/cockroachdb/cockroach/issues/124515
Release note (bug fix): fixed handling in the DSC when columns are included in the STORING() clause of CREATE INDEX. We now check if the column is virtual up front, and properly detect when a column is already handled by an existing index when the column name has UTF-8 characters.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124981)
<!-- Reviewable:end -->

Dedej-Bergin reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124981#-:-NzZi54y5G8oyjFWgza1:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! i left an optional suggestion about adding some detail to the error; curious what you think?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124981#-:-NzZvbwjDQGysryyyUaA:b-fnpe6t)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 618 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124981#-NzZv7py7Omk89DvXQ2c:-NzZv7py7Omk89DvXQ2d:b-sxxmdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca6fa241fbd8eac782d81a5945d5214d87f26ca/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L618)):*
> ```Go
> 		columnName := mustRetrieveColumnNameElem(b, e.TableID, e.ColumnID)
> 		if _, found := columnRefs[columnName.Name]; found {
> 			panic(sqlerrors.NewColumnAlreadyExistsInIndexError(string(n.Name), columnName.Name))
> ```

i noticed that the legacy schema changer also includes an additional DETAIL message when the column is in the primary index, so it returns something like:
```
ERROR: index "index_w0_7_w0_47" already contains column "col\\u000b7ͪ%q_w0_10"
SQLSTATE: 42701
DETAIL: column "col\\u000b7ͪ%q_w0_10" is part of the primary index and therefore implicit in all indexes
```

that extra detail is added as such: https://github.com/cockroachdb/cockroach/blob/970ea323f5d3805688c08a40a28ee2951f24113f/pkg/sql/catalog/tabledesc/structured.go#L746-L750

Perhaps it would be good for the declarative schema changer to also include that detail. i think it would mean we need to conditionally add that message depending on whether it is present in the primary index or not.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: fix bugs with the STORING() clause of CREATE INDEX.

Issue #124511 uncovered two problems with the Declarative Schema Changer
(DSC):

1. If a column name contains a UTF-8 character and is included in the
   STORING() clause, the DSC cannot detect that it duplicates the
   primary key.
2. If a virtual column is included in the STORING() clause, the DSC does
   not catch this upfront, unlike the legacy schema changer, leading to
   an internal error.

The first problem occurs because we build the list of columns referenced
in an index by their names. These names are normalized to handle UTF-8
encoding. However, when checking if a storing column is already
referenced, we weren't using the normalized version of the string,
resulting in a mismatch.

The second problem arises from inadequate checks to detect if a virtual
column is included in the STORING() clause. Adding a check for this
resolves the issue.

Fixes #124511
Fixes #124515
Release note (bug fix): fixed handling in the DSC when columns are
included in the STORING() clause of CREATE INDEX. We now check if the
column is virtual up front, and properly detect when a column is already
handled by an existing index when the column name has UTF-8 characters.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124981)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 618 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124981#-NzZv7py7Omk89DvXQ2c:-NzcQpmo5A0vR-yZqK62:b-egymsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/eca6fa241fbd8eac782d81a5945d5214d87f26ca/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L618)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i noticed that the legacy schema changer also includes an additional DETAIL message when the column is in the primary index, so it returns something like:
```
ERROR: index "index_w0_7_w0_47" already contains column "col\\u000b7ͪ%q_w0_10"
SQLSTATE: 42701
DETAIL: column "col\\u000b7ͪ%q_w0_10" is part of the primary index and therefore implicit in all indexes
```

that extra detail is added as such: https://github.com/cockroachdb/cockroach/blob/970ea323f5d3805688c08a40a28ee2951f24113f/pkg/sql/catalog/tabledesc/structured.go#L746-L750

Perhaps it would be good for the declarative schema changer to also include that detail. i think it would mean we need to conditionally add that message depending on whether it is present in the primary index or not.
</blockquote></details>

If I understand the code correctly, the error you highlighted here is already checking the columns of the primary key. Therefore, I believe I can add the extra detail without any conditions. I'll update the test to cover both cases where we return 'NewColumnAlreadyExistsInIndexError'.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

spilchen commented:
bors r+
